### DISCUSSION TITLE
ImportError when partially packaging sympy

### DISCUSSION THREAD

**Discussion by TomMonkeyMan** [sympy/sympy#27107] [Q&A] [answered]
Hi friends,

I'm working with pyspark and trying to solve symbolic equations using the sympy lib. To use sympy in a distributed spark env, I need to package the dependencies into a `.zip` file and submit them with my script.

Initially, I attempted to package only the `sympy` and `mpmath` directories using the following command:

I installed the sympy with:
```
pip install --target=dependencies sympy
```
and here is the structure of my `dependencies` folder: 
```
dependencies/
├── mpmath/
├── mpmath-1.3.0.dist-info/
├── sympy/
├── sympy-1.13.3.dist-info/
├── __pycache__/
├── bin/
├── share/

```

Initially, I attempted to package only the `sympy` and `mpmath` directories using the following command:
```
zip -r ../dependencies.zip . sympy/* mpmath/*
```

However, when running my PySpark job, I encountered the following error:
```
ImportError: cannot import name 'make_mpc' from 'mpmath' (unknown location)
```

When I packaged the entire `dependencies` directory without excluding any files or folders, like this:
```
zip -r ../dependencies.zip .
```

The error disappeared, and everything worked fine.

I'm trying to understand **why excluding only the `sympy` and `mpmath` directories caused the import error** . The `dependencies` folder also contains other directories like `mpmath-1.3.0.dist-info` , `sympy-1.13.3.dist-info` , `__pycache__` , and `share` . Could the missing metadata or some other files in these folders be necessary for sympy or mpmath to work correctly?

Any insights into what might be causing the issue when only `sympy` and `mpmath` are packaged, and why including all files resolves the problem? I’m particularly interested in understanding if files in `dist-info` , `share` , or `bin` directories are necessary for the library to function properly.

Thanks for your help!

I’m very interested in understanding the root cause of this issue, so I’ve also posted it in another community here: https://discuss.python.org/t/importerror-when-partially-packaging-sympy/64779

**Comment by oscarbenjamin** [comment]
This seems like a pyspark question rather than a sympy one. I don't know anything about pyspark. From your description you are installing sympy in a non-standard way.

I don't think that sympy or mpmath does anything special here. They both just assume that they will be installed to sys.path somewhere.

**Reply by TomMonkeyMan** [reply]
Hi, thanks for the reply! I’m still using sympy as a python package in pyspark. And you’re right, I didn’t install it via pip; instead, I copied the installed folders directly to each node to run it. The issue arises when I only copy the mpmath/* and sympy/* folders—it leads to the import error. However, when I copy the entire installation directory (including metadata and other files), everything works fine.

I’m just curious why copying only those two folders causes the import issue. I understand it’s a non-standard installation method, but I’m not sure what’s missing when I exclude the other files.

**Reply by oscarbenjamin** [reply, marked as answer]
This is still a pyspark question. If you copy the sympy and mpmath directories from site-packages into a directory and add that directory to sys.path then you can import and use both sympy and mpmath just fine. There is nothing special about sympy or mpmath that would prevent that from working.

**Reply by TomMonkeyMan** [reply]
That makes sense. I should try copying the directories and testing it locally first. Thanks for the advice! I’ll do some investigation and get back to you after that.

**Reply by TomMonkeyMan** [reply]
I just did a quick test—created a venv and copied only the sympy and mpmath directories to site-packages, and everything worked fine. It seems to be a pyspark-specific issue. I’ll look more into it. Thanks a lot for the help!

**Reply by TomMonkeyMan** [reply]
re-installed pyspark, which solved the issue.
